Annotate unsupported APIs in .Net core as Obsolete - Reflection area

buyaa-n opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/50529

As part of https://github.com/dotnet/runtime/issues/47228 we are annotating APIs not supported in .Net Core with [Obsolete] attribute

buyaa-n committed:
Annotate unsupported APIs in .Net core as Obsolete -

buyaa-n commented on src/libraries/System.Private.CoreLib/src/System/Reflection/StrongNameKeyPair.cs:
```diff
@@ -29,9 +29,11 @@ public StrongNameKeyPair(byte[] keyPairArray)
         protected StrongNameKeyPair(SerializationInfo info, StreamingContext context) =>
```
This and other serialization related APIs are exposed in the ref maybe we need to annotate them too CC @GrabYourPitchforks

joperezr replied:
Yeah why was that not considered here? Is it because the exception doesn't have the StrongNameSigning message?

buyaa-n replied:
Now i think we should obsolete this StrongNameKeyPair entirely, all its members just throwing, except one constructor `public StrongNameKeyPair(FileStream keyPairFile)` which also doesn't seem like doing anything cc @steveharter

buyaa-n replied:
It is only used here https://github.com/dotnet/runtime/blob/32a86674e4942aeb76cba4d6a689aa207ae237fe/src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs#L203-L207 which doesn't seem used anywhere in runtime

GrabYourPitchforks replied:
Buyaa pointed this out to me offline as well. I agree with her assessment that the entire type can go away. I'd also support removing the code in the ctor and replacing it with a "throw" statement like the other methods do.

joperezr reviewed (approved):
Changes look good, assuming that this has gone through API review and that the breaking change has been approved.

buyaa-n committed:
Disable warning in test

buyaa-n commented:
> assuming that this has gone through API review and that the breaking change has been approved.

There was no API review, but we had an offline discussion that we want to Obsolete APIs not supported in .Net core for all platforms so that developers get warnings at design time instead of getting PNSE at runtime (from that perspective i think it is not breaking change). We filed issues for each area so that area owners handled it appropriately, main thing the area owners need to consider/make sure is the APIs are really not supported on any platform and not intended/planned to get supported in the future. For the APIs in this PR APIs in `Assembly` and `Type` are documented that it is not supported on .NetCore, APIs in `StrongNameKeyPair` not documented but doesn't look like not supported on .Net Core, all of them has no associated issue for future support

buyaa-n committed:
Disable more warnings in test

GrabYourPitchforks commented on src/libraries/System.Private.CoreLib/src/System/Reflection/StrongNameKeyPair.cs:
```diff
@@ -29,9 +29,11 @@ public StrongNameKeyPair(byte[] keyPairArray)
         protected StrongNameKeyPair(SerializationInfo info, StreamingContext context) =>
             throw new PlatformNotSupportedException();
 
+        [Obsolete("Strong-name signing is not supported on this platform.")]
```
Nit: Remove the hyphen so that this reads "Strong name signing". We don't traditionally hyphenate the term "strong name".
(same comment on line 36 below and in the ref assemblies)

GrabYourPitchforks commented on src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs:
```diff
@@ -200,11 +198,8 @@ public AssemblyVersionCompatibility VersionCompatibility
-            set => _strongNameKeyPair = value;
-        }
+        [Obsolete("Strong name signing is not supported on this platform.")]
+        public StrongNameKeyPair? KeyPairr => throw new PlatformNotSupportedException(SR.PlatformNotSupported_StrongNameSigning);
```
We can't get rid of the backing field. The native reflection stack requires it to be present. So this property getter and setter may have to remain present. Let's hold of on obsoleting this property for now and revert it back to what it was.

GrabYourPitchforks replied:
We have a few options for how to respond to this.

1. Obsolete the property, but leave its implementation as-is. You can get and set the property. And if you're doing so while targeting .NET 6, you'll receive a warning.
2. Change the implementation of the property getters / setters to no-op or throw. This maybe an unacceptable breaking change, as it looks like there are [many consumers](https://source.dot.net/System.Private.CoreLib/R/64dc4409ac56c451.html) of this API.
3. Do nothing with the property. Do not mark it obsolete. If the backing type `StrongNameKeyPair` is marked obsolete, then callers will see a warning anyway.

buyaa-n replied:
I am leaning to #1 it is not really used within runtime, and as you mentioned callers will still get the warning from `StrongNameKeyPair` there is no much difference i think. But if you would suggest #3, I will just suppress the warning for the property

GrabYourPitchforks replied:
`#1` seems fine. Consider: After this change, you'll never be able to instantiate a `StrongNameKeyPair` again. So the only thing you'll ever be able to do is set _KeyPair_ to null or retrieve null from the _KeyPair_ property. Totally fine IMO to mark the properties as obsolete since they're completely useless at that point.

I suppose we could also remove the native reflection support for this as well at the same time, but that's up to you.

Buyaa, what do you think about defining the property as `get => null; set => { };`? That would allow you to delete the backing field as you had originally wanted to do, assuming you also delete the following lines from the native reflection stack at the same time.

https://github.com/dotnet/runtime/blob/927b1c54956ddb031a2e1a3eddb94ccc16004c27/src/coreclr/vm/object.h#L1505

https://github.com/dotnet/runtime/blob/927b1c54956ddb031a2e1a3eddb94ccc16004c27/src/coreclr/vm/object.h#L1518

buyaa-n replied:
> I suppose we could also remove the native reflection support for this as well at the same time, but that's up to you.

From my search seems that is not referenced anywhere, so probably we could remove that

> Buyaa, what do you think about defining the property as get => null; set => { };? That would allow you to delete the backing field as you had originally wanted to do, assuming you also delete the following lines from the native reflection stack at the same time.

The only concern is those references from MSBuild you mentioned in `#2`, i saw them before updating to `throw PNSE`, but thought those are old/stale probably unused/unreachable code as the type itself is not doing anyways, so i was expecting MSBuild would need some cleanup, but my expectation might be wrong, maybe we should talk with somebody from MSBuild?

GrabYourPitchforks replied:
> maybe we should talk with somebody from MSBuild?

A good idea, certainly! :)

buyaa-n replied:
I am not sure whom to talk to, so i created an [issue](https://github.com/dotnet/msbuild/issues/6343) in MsBuild repo

GrabYourPitchforks commented:
According to source.dot.net, there are only two instantiations of `StrongNameKeyPair` throughout the entirety of the indexed code bases, and both of them would have thrown `PNSE` anyway. So for these callers, changing the `StrongNameKeyPair` ctor is not a breaking change.

https://github.com/dotnet/msbuild/blob/9bcc06cbe19ae2482ab18eab90a82fd079b26897/src/Tasks/StrongNameUtils.cs#L68-L84
https://github.com/dotnet/msbuild/blob/9bcc06cbe19ae2482ab18eab90a82fd079b26897/src/Tasks/StrongNameUtils.cs#L99-L115

buyaa-n committed:
Clean up all StrongNameKeyPair references

buyaa-n committed:
Merge conflict

buyaa-n committed:
Update obsoletion warning ID

buyaa-n committed:
Add the the obsoletion into the list-of-diagnostics.md

buyaa-n commented:
@GrabYourPitchforks 
1. MsBuild folks replied that If they reference it in .NET 5/6, they would ifdef it out. 
2. As we discussed offline obsoleted the `StrongNameKeyPair` type with separate diagnostic ID and removed the backing field in `AssemblyName` type, also removed all references in managed and native source

The PR is now ready for review, please take another look

jeffhandley commented on docs/project/list-of-diagnostics.md:
```diff
@@ -28,6 +28,7 @@ Currently the identifiers `SYSLIB0001` through `SYSLIB0999` are carved out for o
 |  __`SYSLIB0013`__ | Uri.EscapeUriString can corrupt the Uri string in some cases. Consider using Uri.EscapeDataString for query string components instead. |
 |  __`SYSLIB0015`__ | DisablePrivateReflectionAttribute has no effect in .NET 6.0+ applications. |
 |  __`SYSLIB0016`__ | Use the Graphics.GetContextInfo overloads that accept arguments for better performance and fewer allocations. |
+|  __`SYSLIB0017`__ | Strong name signing is not supported on this platform.|
```
We can probably omit the "on this platform" portion of the message. Similar to the others, we might want to say the API throws `PlatformNotSupportedException`.

buyaa-n replied:
The message copied from the resource string passed in PNSE constructor, i can update

buyaa-n commented on src/libraries/System.Private.CoreLib/src/System/Reflection/StrongNameKeyPair.cs:
```diff
@@ -6,25 +6,14 @@
 
 namespace System.Reflection
 {
+    [Obsolete(Obsoletions.StrongNameKeyPairMessage, DiagnosticId = Obsoletions.StrongNameKeyPairDiagId)]
```
I saw that but didn't put that because there is no doc corresponding to this obsoletion and doesn't seem to need to have one as  there is nothing really put into the doc. Compared to the other obsoletions it doesn't have any implementation and no work around/alternative (as far asi know). So there is nothing to put in the doc expect the message that it is not supported in .NetCore, i am planning to add a note in the existing API doc, but just didn't felt we need separate doc for this obsoletion, please let me know if we still want the doc url

jeffhandley replied:
I do think we want to have docs for each obsoletion for discoverability, even if the doc is short. @GrabYourPitchforks -- do you agree?

buyaa-n replied:
If not the same id for all throwing PNSE I guess we would at least group them by the relevance/type and give separate ids for each group. Then for this PR only one ID will be added for `ReflectionOnly loading`

jeffhandley replied:
> Then for this PR only one ID will be added for `ReflectionOnly loading`

That sounds good and matches the pattern we've been following with the other obsoletions.

buyaa-n replied:
Updated all, thanks!

buyaa-n commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs:
```diff
@@ -372,10 +372,13 @@ public static Assembly LoadFrom(string assemblyFile, byte[]? hashValue, Assembly
         [RequiresUnreferencedCode("Types and members the loaded module depends on might be removed")]
         public virtual Module LoadModule(string moduleName, byte[]? rawModule, byte[]? rawSymbolStore) { throw NotImplemented.ByDesign; }
 
+        [Obsolete("ReflectionOnly loading is not supported on this platform.")]
```
I am not sure if we want to add separate obsoletion for all this, we might want to have on generic diagnostic id for all the obsoletions for APIs not supported in .Net core

jeffhandley replied:
I do think we want to have docs for each obsoletion for discoverability, even if the doc is short. @GrabYourPitchforks -- do you agree?

jeffhandley replied:
Oops; that comment was more appropriate on the [previous thread](https://github.com/dotnet/runtime/pull/50941#discussion_r616831379).

@GrabYourPitchforks / @gewarren What do you think about having a generalized obsoletion document for APIs that were only supported on .NET Framework?

buyaa-n replied:
Probably we use the same diagnostic id but different messages

gewarren replied:
> generalized obsoletion document

We do have this page for APIs that throw PNSE: https://docs.microsoft.com/en-us/dotnet/core/compatibility/unsupported-apis#systemreflection. And this page for SYSLIB obsoletions (which looks like it needs to be updated?): https://docs.microsoft.com/en-us/dotnet/core/compatibility/syslib-obsoletions. Would one of these pages work?

buyaa-n replied:
> We do have this page for APIs that throw PNSE: https://docs.microsoft.com/en-us/dotnet/core/compatibility/unsupported-apis#systemreflection. 

That shows like 100+ APIs, do we want different diagnostic ID for each?

GrabYourPitchforks commented on src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs:
```diff
@@ -200,10 +198,11 @@ public AssemblyVersionCompatibility VersionCompatibility
             set => _versionCompatibility = value;
         }
 
+        [Obsolete(Obsoletions.StrongNameKeyPairMessage, DiagnosticId = Obsoletions.StrongNameKeyPairDiagId)]
```
Don't forget _UrlFormat_ here!

GrabYourPitchforks reviewed (approved):
A missing _UrlFormat_, but otherwise LGTM. Thanks!

buyaa-n committed:
Apply feedbacks

jeffhandley reviewed (approved):
It looks like there's a check failing in `src/mono/System.Private.CoreLib/src/System/RuntimeTypeHandle.cs` where another pragma will be needed. But I preemptively approve based on that getting resolved.

buyaa-n committed:
Suppress warning in mono

buyaa-n commented on src/mono/System.Private.CoreLib/src/System/RuntimeTypeHandle.cs:
```diff
@@ -308,7 +308,9 @@ internal static bool IsTypeDefinition(RuntimeType type)
                 try
                 {
+#pragma warning disable SYSLIB0018 // ReflectionOnly loading is not supported and throws PlatformNotSupportedException.
                     a = Assembly.ReflectionOnlyLoad(an);
```
Not sure if mono is used in lower versions, does not look like it is, if that is the case we probably should just throw PNSE in this `reflectionOnly` case cc @marek-safar

marek-safar replied:
Why do you think it's Mono specific? I think we should bubble up this information not hide it if always throws

buyaa-n replied:
> Why do you think it's Mono specific? 

It didn't warn in CoreClr corelib, I don't even see this method having ` bool reflectionOnly` parameter in CoreClr

>  I think we should bubble up this information not hide it if always throws

It is not throwing unconditionally, as you know we cannot conditionally Obsolete an API

Anyway, I see this API always called with a value of `false` for `bool reflectionOnly` parameter, we might want to remove this parameter from this method, probably with a different PR

jeffhandley replied:
> I see this API always called with a value of `false` for `bool reflectionOnly` parameter, we might want to remove this parameter and related code section from this method, probably with a different PR

Sounds good to me.

buyaa-n replied:
Created an issue for this https://github.com/dotnet/runtime/issues/51838

buyaa-n committed:
Merge conflicts

jeffhandley commented on src/libraries/Directory.Build.targets:
```diff
@@ -17,7 +17,7 @@
     <NoWarn Condition="'$(GeneratePlatformNotSupportedAssembly)' == 'true' or '$(GeneratePlatformNotSupportedAssemblyMessage)' != ''">$(NoWarn);nullable;CA1052</NoWarn>
     <!-- Ignore Obsolete errors within the generated shims that type-forward types.
          SYSLIB0003 is for CAS. SYSLIB0004 is the obsoletion of CER. -->
```
```suggestion
         SYSLIB0003: Code Access Security (CAS). SYSLIB0004: Constrained Execution Region (CER). SYSLIB0017: Strong name signing.  -->
```

buyaa-n committed:
Update comment

RussKie commented:
What course of action should the downstream consumers affected by this change take? E.g. [dotnet/winforms](https://github.com/dotnet/winforms/pull/4848) is currently broken because `StrongNameKeyPair` is a part of our public API.

* Do we remove the public API? 
* Do we mark it as obsolete? If so, what obsoletion message should we put, as the runtime obsoletion strings aren't available to us.

/cc: @terrajobst

terrajobst commented:
You could also just suppress the obsoletion. I doubt anyone would care :-)

RussKie commented:
I didn't consider it, that's an option too :)
Thanks!
